### ISSUE TITLE
Explicit support for auto closable buffers

### ISSUE THREAD

**Issue by brokenpylons** [neovim/neovim#11752]
Not all buffers are used for displaying files (REPLs, NERDtree, Mundo...), so it would be nice if these could be marked as such, so they get closed automatically when neovim is closed. There are workarounds for this:

https://github.com/preservim/nerdtree/issues/21
https://yous.be/2014/11/30/automatically-quit-vim-if-actual-files-are-closed/

However these solutions are prone to breaking and each such buffer must be supported individually.

**Comment by justinmk** [comment]
The linked issues are too long, please summarize the important parts here. 

> it would be nice if these could be marked as such

How is this different than `:set buftype=nofile` ?

> when neovim is closed.

Closed _how_? Say exactly what steps reproduce the behavior and what you expect to happen.

**Comment by brokenpylons** [comment]
Sure, here is a use case using the Mundo plugin as an example:

1. open a file (```nvim test```)

![issue1](https://user-images.githubusercontent.com/21318936/73135664-e9a3c780-4044-11ea-91ca-f8353568a7cd.png)

2. open mundo and switch back to the file (```:MundoToggle```, ```<c-w>l```)

![issue2](https://user-images.githubusercontent.com/21318936/73135683-22dc3780-4045-11ea-8cb9-f2aace25a752.png)

3. close the file (```:q```)

![issue3](https://user-images.githubusercontent.com/21318936/73135684-2d96cc80-4045-11ea-8c3d-ff71e42d10d2.png)

Actual behavior: Neovim doesn't close and only the Mundo window is left
Expected behavior: Neovim closes

> How is this different than ```:set buftype=nofile```

I want the window to close also in the case when ```:q``` is used.

So, the way I see it, an ideal solution would be that there would be something like```:set bufftype=autoclose```, which would mean the buffer would generally behave just like ```:set bufftype=nofile```, but in case only buffers of these type are left neovim would just close. 

Why isn't ```:qall``` good enough? These windows are just user interfaces for plugins, there is no good reason for them to stay open after all buffers used for editing text were closed, so the fact that they stay open is just an annoyance.

Why aren't existing solutions good enough? Because these buffers aren't marked in a consistent and unique way. You can generally infer that some buffer belongs to such and such plugin from it's name, but this a very fragile solution, the name doesn't guarantee anything and you would need to handle each plugin separately, which is not convenient. You could also use the fact that the buffer is marked as  ```:set bufftype=nofile```, but that doesn't tell you whether it is an user interface or just a scratch buffer and you don't want scratch buffers to close automatically in the same way.

Also sorry for not providing complete information, I though the problem was common enough, I hope that this clears things up :)

**Comment by justinmk** [comment]
> Because these buffers aren't marked in a consistent and unique way.

Introducing yet another 'buftype' (or other flag) won't help that. 

> You could also use the fact that the buffer is marked as `:set bufftype=nofile`, but that doesn't tell you whether it is an user interface or just a scratch buffer and you don't want scratch buffers to close automatically in the same way.

`bufftype=nofile` by definition means that you don't care about the buffer and it must not provoke a confirmation prompt.

Anyways you can make a trivial plugin that does what you want:

    :au QuitPre * qall

then `:q` will close all windows and exit Nvim. Isn't difficult to then check for 'buftype' or whatever else you are expecting.

**Comment by brokenpylons** [comment]
> Introducing yet another 'buftype' (or other flag) won't help that.

Sure, not immediately. But the situation is the way it is, because there is currently no standard way to do this. Also if you scroll over the issue I linked in the original post you will see it was introduced in 2010 and the last comment is from 2019, so it's a constant problem.

Also to be clear I'm not looking for a quick fix, I already solved it for my use case (albeit not as elegantly as your one liner), but as I said I would be nice if this was supported in the editor, so it would "just work" by default.
